File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    int goodNodes(TreeNode *root)
    {
        if (root == nullptr)
        {
            return 0;
        }

        queue<pair<TreeNode *, int>> tree;

        tree.push(make_pair(root, root->val));
        int res = 0;

        while (!tree.empty())
        {
            pair<TreeNode *, int> curr = tree.front();
            tree.pop();
            if (curr.first->left != nullptr)
                tree.push(make_pair(curr.first->left, max(curr.first->val, curr.second)));
            if (curr.first->right != nullptr)
                tree.push(make_pair(curr.first->right, max(curr.first->val, curr.second)));
            if (curr.first->val >= curr.second)
            {
                res += 1;
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/subsets-ii/submission-1.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []

        def backtrack(i: int, curr: List[int]):
            if i == len(nums):
                res.append(curr.copy())
                return

            curr.append(nums[i])
            backtrack(i + 1, curr)
            curr_val = curr.pop()
            while i < len(nums) and nums[i] == curr_val:
                i += 1
            backtrack(i, curr)
        backtrack(0, [])
        return res